python/seldon_core/metrics.py

```diff
@@ -1,10 +1,115 @@
+import logging
+import os
+
+
+logger = logging.getLogger(__name__)
+
 
 COUNTER = "COUNTER"
 GAUGE = "GAUGE"
 TIMER = "TIMER"
 
 
+def generate_metrics(metrics):
+    myregistry = CollectorRegistry()
+    myregistry.register(metrics)
+    return (
+        exposition.generate_latest(myregistry).decode("utf-8"),
+        exposition.CONTENT_TYPE_LATEST,
+    )
+
+
+BINS = [0] + list(np.logspace(-3, np.log10(30), 50)) + [np.inf]
+LABELS = ["worker-id", "model", "image"]
+
+my_labels = {"model": "latest", "image": "my-image"}
+
+
+def update_hist(x, vals, sumv):
+    hist = np.histogram([x], BINS)[0]
+    vals = list(np.array(vals) + hist)
+    return vals, sumv + x
+
+
+def expose_gauge(name, value, labels):
+    metric = GaugeMetricFamily(name, "", labels=LABELS)
+    metric.add_metric(labels, value)
+    return metric
+
+
+def expose_counter(name, value, labels):
+    metric = CounterMetricFamily(name, "", labels=LABELS)
+    metric.add_metric(labels, value)
+    return metric
+
+
+def expose_histogram(name, value, labels):
+    vals, sumv = value
+    buckets = [[floatToGoString(b), v] for v, b in zip(np.cumsum(vals), BINS[1:])]
+
+    metric = HistogramMetricFamily(name, "", labels=LABELS)
+    metric.add_metric(labels, buckets, sum_value=sumv)
+    return metric
+
+
+class SeldonMetrics:
+    """Class to manage custom metrics stored in shared memory."""
+
+    def __init__(self, worker_id_func=os.getpid):
+        self.manager = Manager()
+        self.data = self.manager.dict()
+        self.worker_id_func = worker_id_func
+
+    def __del__(self):
+        self.manager.shutdown()
+
+    def update(self, custom_metrics):
+        data = self.data.get(self.worker_id_func(), {})
+
+        for metrics in custom_metrics:
+            key = metrics["type"], metrics["key"]
+            if metrics["type"] == "COUNTER":
+                value = data.get(key, 0)
+                data[key] = value + metrics["value"]
+            elif metrics["type"] == "TIMER":
+                vals, sumv = data.get(key, (list(np.zeros(len(BINS) - 1)), 0))
+                # Dividing by 1000 because unit is milliseconds
+                data[key] = update_hist(metrics["value"] / 1000, vals, sumv)
+            else:
+                data[key] = metrics["value"]
+
+        self.data[self.worker_id_func()] = data
+
+    def collect(self):
+        data = dict(self.data)
+
+        for worker, metrics in data.items():
+            labels = [str(worker), my_labels["model"], my_labels["image"]]
+            for (item_type, item_name), item_value in metrics.items():
+                if item_type == "GAUGE":
+                    yield expose_gauge(item_name, item_value, labels)
+                elif item_type == "COUNTER":
+                    yield expose_counter(item_name, item_value, labels)
+                elif item_type == "TIMER":
+                    yield expose_histogram(item_name, item_value, labels)
+                else:
+                    continue
```

Comment:
Doesn't seem required as if it doesn't match the next one it will continue anyways

Reply:
Indeed. Removed redundant block.